### ISSUE TITLE
The `:matches(…)` pseudo‑class has been renamed to `:is(…)`

### ISSUE THREAD

**Issue by ExE-Boss** [dperini/nwsapi#28]
The only difference is that `:is(…)` follows `@media` query‑like invalidation:
- https://github.com/w3c/csswg-drafts/issues/3264#issuecomment-440752606
- https://github.com/w3c/csswg-drafts/issues/3264#issuecomment-467190906

**Comment by dperini** [comment]
@ExE-Boss
thank you for your time and suggestion, I always appreciate contributions.
We are fully aware of these API name changes in current drafts but they are still `drafts`.

I have been discussing this API name change with the maintainers of `jsdom` and it was decided to let this as is, thus follow what is in real browser as this is the main objective of `jsdom` (which makes up 90% usage of the `nwsapi` project).

I do like the new Selectors API name and in regard to CSS I like the new semantic flow that `:is()` gives coupled with the `:not()` selector and I am prepared for the change, just waiting for things to settle. In `nwsapi` code It is just a matter of adding `is()` as the default API name and change `matches()` to be an alias kept for backward compatibility.

As for other reasons to wait implement this change I can say that none of the browsers I can test
has implemented this new API name change yet. The name of this method has remained unchanged so every browser implementation we know is still using `matches(...)`.

For the above reasons I do not agree about this being an issue in the current code.

NOTE: a new version with code cleanup and some extra speed improvements is on its way. now.

**Comment by ExE-Boss** [comment]
> In `nwsapi` code It is just a matter of adding `is()` as the default API name and change `matches()` to be an alias kept for backward compatibility.

If you only do that, then `:is(foo, :maybe-unsupported)` will throw an error instead of matching all `<foo>` elements.

I don’t actually know where to insert the code for the better `@media`‑style error handling within `nwsapi`’s implementation.

**Comment by dperini** [comment]
@ExE-Boss 
I don't know yet if the final "Recommendations" will include those rules and, given the history of tentative changes that were dropped in the past I am a bit reluctant to start implementing in advance. Even if I did that the change you suggest, it will probably be against `jsdom` rules of not being implemented in browser. I would have to ask @domenic and @Zirro to confirm or negate this position.

It seems to me you are saying that:

    document.documentElement.matches('html', ':maybe-unsupported');
would return 'false' ... instead in my tests it does return 'true' ... if I did understand what you are saying. Please correct me if I am wrong or if I did misunderstand you.

**Comment by ExE-Boss** [comment]
That would return `true`, because `arguments` is:
```jsonc
[
	"html", // only the first argument contains a selector
	":maybe-unsupported"
]
```

What you probably meant was:
```js
document.documentElement.matches('html, :maybe-unsupported');
```
Which will throw a `SyntaxError` if `:maybe‑unsupported` isn’t supported.

---

This should however return `true`:
```js
document.documentElement.matches(':is(html, :maybe-unsupported)');
```

**Comment by dperini** [comment]
@ExE-Boss 
my bad ... You are correct `matches(...)` only accepts one selector string.
So when changing the name of the API I should have a look to that difference.

Unfortunately I don't have a platform where I can test this behaviour yet.
If you happen to have one to suggest I could start exercising some adjustment.

Thank you for the head up on this W3C functionality change of this API method.

**Comment by ExE-Boss** [comment]
The functionality of the API method isn’t changing.

**Comment by danburzo** [comment]
Firefox 78 has landed support for the [`:is()`](https://www.w3.org/TR/selectors-4/#matches) and [`:where()`](https://www.w3.org/TR/selectors-4/#zero-matches) pseudo-classes, and there are a few ([1](https://github.com/web-platform-tests/wpt/blob/master/css/selectors/is-nested.html), [2](https://github.com/web-platform-tests/wpt/blob/master/css/selectors/is-specificity.html), [3](https://github.com/web-platform-tests/wpt/blob/master/css/selectors/is-where-parsing.html)) web platform tests for them. The upcoming Safari 14 presumably enable them by default, and Chrome is still [working on them](https://bugs.chromium.org/p/chromium/issues/detail?id=808904). Would it be a good time to discuss their implementation in nwsapi?

(I found this thread after logging [an issue on `jsdom`](https://github.com/jsdom/jsdom/issues/3013) about the `:matches()` pseudo-class.)

**Comment by dperini** [comment]
@danburzo 
yes, changing that naming was the plan for "nwsapi" too.
Was just waiting for browsers to show up their implementation.

**Comment by danburzo** [comment]
Awesome! Please let me know if there's anything I can help with.

**Comment by dperini** [comment]
@danburzo 
yes ... thank you for offering help. It is appreciated.
Testing the changes I will commit and executing tests would be great.
Also checking the problem with the "matches() > is()" renaming outlined by @ExE-Boss would help.
I already made this change and the suggested "nodeName > localName" replacement as suggested in issue #37 
I believe tomorrow I will be pushing these two commits.

**Comment by danburzo** [comment]
I'll be happy to test the changes. 

In regards to the problem raised by @ExE-Boss, if we're referring to the `:is()` and `:where()` pseudo-classes being more permissive with invalid selectors, unfortunately it seems Firefox has shipped without it. 

In fact, in Firefox this sequence:

```js
document.querySelector('div:is(div, :bogus)');
```

...throws an error instead of ignoring `:bogus`. I'm not sure where the proposal stands in light of this, but I'm going to ask on the CSSWG thread directly.

**Comment by dperini** [comment]
@danburzo @ExE-Boss 
I pushed a commit for this change.
I will appreciate a review on it's functionality compared to that of recent Firefox as it is the only browser that already made the change. In this case the objective is mimicking the same functionality found on current browsers implementation rather than following the specification to the letter. As soon as other browsers implement this, we can level up our implementation.
Thank you.

**Comment by ExE-Boss** [comment]
@dperini
> I pushed a commit for this change.

No, you didn’t, as the `:is(…)` pseudo‑class is still unrecognised: <https://github.com/dperini/nwsapi/blob/44a937dbeb243459f13ae7d36f1ed1413884b2c4/src/nwsapi.js#L79><https://github.com/dperini/nwsapi/blob/44a937dbeb243459f13ae7d36f1ed1413884b2c4/src/nwsapi.js#L1006-L1008>

**Comment by dperini** [comment]
@danburzo 
my bad, sorry ... there was a big misunderstanding on my side.
I thought the API name was changed from `.macthes()` to `.is()` while I see now only the CSS pseudo-class names have changed.
I removed the wrong commit and I will prepare for the right one.

**Comment by dperini** [comment]
@danburzo 
about nesting :not() pseudo-classes ....
>>> (was there a specific reason why it was prohibited in the existing codebase?);
Yes the reason is here and possibly still valid:
https://developer.mozilla.org/en-US/docs/Web/CSS/:not#Description

**Comment by dperini** [comment]
@danburzo 
yes, your changes to the code seems to be correct.
I believe we should leave ':matches' as an alias for backward compatibility so the RegExp would be:

+    logicalsel: '(is|where|matches|not)\\x28\\s?([^()]*|[^\\x28]*\\x28[^\\x29]*\\x29)\\s?(?:\\x29|$)',

and similarly, later:

+                case 'is':
+                case 'where':
+                case 'matches':

I don't think we should remove `:matches` completely. Should we ?

**Comment by danburzo** [comment]
The [Selectors Level 4](https://www.w3.org/TR/selectors-4) specification has relaxed some of these requirements:

> 19.4. Changes since the 23 August 2012 Working Draft
> Significant changes since the 23 August 2012 Working Draft include:
> * Released some restrictions on :matches() and :not(). 

That means `:is()`, `:where()` and `:not()` pseudo-classes can be combined at will. 

The behavior in current browsers:

* __Firefox:__ Pseudo-classes can be combined, but still considers invalid selectors containing _nested `:not`s_. 
* __Safari Technology Preview:__ Pseudo-classes can be used in any combination, including `:not(:not())`. 

If it doesn't affect any other part of the code, I would say not impose any limitation on nested `:not()`s, since that's the intention in the Level 4 spec and it makes the implementation simpler.

**Comment by danburzo** [comment]
> I don't think we should remove :matches completely. Should we ?

Agreed, it makes sense to keep `:matches` around.

**Comment by domenic** [comment]
It would be good to remove `:matches()`; jsdom at least would prefer to be spec-compliant, and `:matches()` does not appear in any spec.

**Comment by dperini** [comment]
@domenic 
I could remove it, but what about Chrome, Safari, IE et all ?
From what I understand ':matches()' pseudo-class is still the only one valid there ! Or am I missing something ?

**Comment by danburzo** [comment]
[According to caniuse](https://caniuse.com/#feat=css-matches-pseudo):

* Chrome never had support for `:matches()`, only for `:-webkit-any()`
* IE does not have support for any of these classes
* Safari up to 13.1 used `:-webkit-any()` and `:-matches()` (My current version, 13.1.1, uses `:matches()`)

So, on second thought, there doesn't appear there was much backwards compatibility we're forfeiting here. Furthermore, due to how `:matches()` was implemented, it did not work correctly for a list of selectors (`:matches(h1, h2)` would try to find elements that are both `h1` and `h2`).

**Comment by ExE-Boss** [comment]
@domenic

> It would be good to remove `:matches()`; jsdom at least would prefer to be spec-compliant, and `:matches()` does not appear in any spec.

According to [Selectors 4](https://drafts.csswg.org/selectors-4/#matches), <code>[:matches()][selectordef-matches]</code> is a valid legacy alias:

> As previous drafts of this specification used the name <code>[:matches()][selectordef-matches]</code> for this pseudo-class, UAs may additionally implement this obsolete name as an alias for <code>[:is()][selectordef-is]</code> if needed for backwards-compatibility.

[selectordef-is]: https://drafts.csswg.org/selectors-4/#matches-pseudo
[selectordef-matches]: https://drafts.csswg.org/selectors-4/#selectordef-matches

**Comment by vincenthongzy** [comment]
Hey there,  could we close this issue since its already fixed and released in `v2.2.1` ? I was tracking some problems related to the `:is` pseudo-class being thrown as Syntax Error in my jest tests, then I stumbled upon this issue. Luckily I checked the commit history, otherwise I would be blocked for a longer period of time.
